### PR TITLE
RUMM-416 FileReader - fix the concurrency issue

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

This PR fixes the concurrency issue we were experiencing in the FileReader. In short this class was using 2 sets to mark files as *in progress* or *sent* and only one of these 2 collections was synchronised. Furthermore there was a multithreaded access to this file (from the WorkManager and from the default DataUploadScheduler) and the application was seldom crashing with a ConcurrentModificationAccess.

### Motivation

Opened issue #234  on our Github repository 

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [x] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [x] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/data/file/FileReaderTest.kt [hunk 1]
```diff
@@ -91,12 +95,13 @@ internal class FileReaderTest {
...
         file.writeText(data)
-        whenever(mockOrchestrator.getReadableFile(any())) doReturn null
-        whenever(mockOrchestrator.getReadableFile(emptySet())) doReturn file
+        whenever(mockOrchestrator.getReadableFile(mutableSetOf())) doReturn file
+        whenever(mockOrchestrator.getReadableFile(mutableSetOf(fileName))) doReturn null
 
         val firstBatch = testedReader.readNextBatch()
         val secondBatch = testedReader.readNextBatch()
         checkNotNull(firstBatch)
+        assertThat(secondBatch).isNull()
```
Comment (xgouchet) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/data/file/FileReaderTest.kt:104]: There's already a check on that below (line 110)

File: dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/data/file/FileReaderTest.kt [hunk 2]
```diff
@@ -193,15 +198,17 @@ internal class FileReaderTest {
     ) {
         // given
         val fileName = forge.anAlphabeticalString()
-        generateFile(fileName)
+        val file: File = generateFile(fileName)
+        whenever(mockOrchestrator.getReadableFile(any())).thenReturn(file)
+        testedReader.readNextBatch()
 
-        // when
+        // f
```
Comment (xgouchet) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/data/file/FileReaderTest.kt:205]: ??

File: dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/domain/FilePersistenceStrategyTest.kt [hunk 3]
```diff
@@ -267,21 +269,27 @@ internal abstract class FilePersistenceStrategyTest<T : Any>(
     }
 
     @Test
-    fun `read returns null when 1st batch is already sent but file still present`(
+    fun `reads null when batch already sent but the other thread is still trying to delete this`(
```
Comment (xgouchet) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/domain/FilePersistenceStrategyTest.kt:272]: I think you should still keep the old test, and add this one too.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/data/file/FileReaderTest.kt (hunk 2 above)
Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/data/file/FileReaderTest.kt:205]: :) don't what I wanted to say there

File: dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/domain/FilePersistenceStrategyTest.kt (hunk 3 above)
Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/domain/FilePersistenceStrategyTest.kt:272]: Well....that's the thing...the first test is not relevant anymore because now we clean the `lockedFiles` if the file is removed from the system. I can add a test maybe to return null if the file could not be deleted.
